### ISSUE TITLE
Pattern matching without K seemingly illogical for the inductive family of squares

### ISSUE THREAD

**Issue by guillaumebrunerie** [agda/agda#3034]
Consider the code below. I define the type of squares filling a given boundary as an inductive family, then I try to pattern match on `ids`, where three sides of it are `idp` and the fourth one is free.
It turns out that `J2` and `J4` type check, but `J1` and `J3` do not.

Is it normal that `J2` and `J4` type check? It’s consistent with `--without-K`, but it is non trivial to prove, so it is potentially surprising that it works at all.
If yes, is it normal that `J1` and `J3` do not type check? The difference is that in the definition of `Square`, the first and third sides involve the variable `a`, which is the only parameter (the others are indices), but I’m not sure why it makes a difference. I tried to put `a` as an index instead of a parameter, but then none of them work.

Ideally I would like them all to work, but I don’t know if the theory allows it.

```agda
{-# OPTIONS --without-K #-}

data _==_ {A : Set} (a : A) : A → Set where
  idp : a == a

data Square {A : Set} {a : A} : {b c d : A} (p : a == b) (q : c == d) (r : a == c) (s : b == d) → Set where
  ids : Square {a = a} idp idp idp idp

-- J1 : {A : Set} {a : A} {p : a == a} → Square p idp idp idp → Set
-- J1 ids = {!!}

J2 : {A : Set} {a : A} {p : a == a} → Square idp p idp idp → Set
J2 ids = {!!}

-- J3 : {A : Set} {a : A} {p : a == a} → Square idp idp p idp → Set
-- J3 ids = {!!}

J4 : {A : Set} {a : A} {p : a == a} → Square idp idp idp p → Set
J4 ids = {!!}
```
Error message for `J1`:
```
I'm not sure if there should be a case for the constructor ids,
because I get stuck when trying to solve the following unification
problems (inferred index ≟ expected index):
  {_} ≟ {_}
  {_} ≟ {_}
  {_} ≟ {_}
  idp ≟ p
  idp ≟ idp
  idp ≟ idp
  idp ≟ idp
Possible reason why unification failed:
  Cannot eliminate reflexive equation a = a of type A because K has
  been disabled.
when checking that the pattern ids has type Square p idp idp idp
```

**Comment by jespercockx** [comment]
`J2` and `J4` are definitely working as intended: this is making full use of general [higher-dimensional unification](https://jesper.sikanda.be/files/higher-dimensional-unification.pdf) to solve the `idp ≟ idp` equations based on their (heterogeneous) types, which are unfortunately not shown in the error message.

The reason why the equations `idp ≟ idp` are not solved for `J1` and `J3` is because the type of (at least one of) the equations is 'too heterogeneous': they're heterogeneous in their parameter (i.e. the lhs of the identity type), which the current implementation of the injectivity rule is not ready for. Indeed, if you change to the two-index version of the identity type then all four definitions pass.

Up until now we conservatively assumed that if the parameters were heterogeneous, there would always be another equation that could be solved first to make them homogeneous, but this example shows that's not the case. One fairly simple solution would be to treat parameters the same as indices for the purpose of higher-dimensional unification, by making them part of the constructor telescope. This would lead in many cases to some extra trivial equations (where one side is a variable), but other than that I don't see any big downsides. We could probably do the same for the conflict and cycle rules to make them work in case of heterogeneous parameters as well.

**Comment by guillaumebrunerie** [comment]
> Indeed, if you change to the two-index version of the identity type then all four definitions pass.

Awesome, I had tried to put `a` as an index for the definition of `Square`, but I didn’t think of doing the same for the identity type. Is there any drawback in doing that?
Anyway, I’ll try it in my development to see if I have any issue.

**Comment by jespercockx** [comment]
I tried to fix this a while back by treating datatype parameters as indices for the purpose of higher-dimensional unification, but this results in some added dot patterns for eta-unit types which the forcing translation doesn't like (it throws an internal error). Maybe we should rather have a separate pass for eta-expanding patterns to make the forcing analysis happy...

**Comment by jespercockx** [comment]
Blocking the fix of #4163 (see https://github.com/agda/agda/pull/4346)

**Comment by jespercockx** [comment]
Maybe my old fix will now work since the forcing translation has been fixed.
